In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [18]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px

In [22]:
df = pd.read_csv("/content/drive/MyDrive/Collab Datasets/DA Exp 6/UniversalBank.csv")
df = df.drop(columns=['ID', 'ZIP Code'], axis=1)
df

# Id and ZIP code columns are removed as they are of no use

Age  Experience  Income  Family  CCAvg  Education  Mortgage  \
0      25           1      49       4    1.6          1         0   
1      45          19      34       3    1.5          1         0   
2      39          15      11       1    1.0          1         0   
3      35           9     100       1    2.7          2         0   
4      35           8      45       4    1.0          2         0   
...   ...         ...     ...     ...    ...        ...       ...   
4995   29           3      40       1    1.9          3         0   
4996   30           4      15       4    0.4          1        85   
4997   63          39      24       2    0.3          3         0   
4998   65          40      49       3    0.5          2         0   
4999   28           4      83       3    0.8          1         0   

      Personal Loan  Securities Account  CD Account  Online  CreditCard  
0                 0                   1           0       0           0  
1                 0                   1           0       0           0  
2                 0                   0           0       0           0  
3                 0                   0           0       0           0  
4                 0                   0           0       0           1  
...             ...                 ...         ...     ...         ...  
4995              0                   0           0       1           0  
4996              0                   0           0       1           0  
4997              0                   0           0       0           0  
4998              0                   0           0       1           0  
4999              0                   0           0       1           1  

[5000 rows x 12 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Age                 5000 non-null   int64  
 1   Experience          5000 non-null   int64  
 2   Income              5000 non-null   int64  
 3   Family              5000 non-null   int64  
 4   CCAvg               5000 non-null   float64
 5   Education           5000 non-null   int64  
 6   Mortgage            5000 non-null   int64  
 7   Personal Loan       5000 non-null   int64  
 8   Securities Account  5000 non-null   int64  
 9   CD Account          5000 non-null   int64  
 10  Online              5000 non-null   int64  
 11  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(11)
memory usage: 468.9 KB


In [26]:
fig = px.scatter(df, x='CCAvg', y='Age', color='Personal Loan')
fig.show()

In [6]:
df1 = df.copy()

In [7]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
df1['Age'] = scaler.fit_transform(df1[['Age']])
df1['Experience'] = scaler.fit_transform(df1[['Experience']])
df1['Income'] = scaler.fit_transform(df1[['Income']])
df1['Mortgage'] = scaler.fit_transform(df1[['Mortgage']])
df1['Family'] = scaler.fit_transform(df1[['Family']])
df1['CCAvg'] = scaler.fit_transform(df1[['CCAvg']])
df1['Education'] = scaler.fit_transform(df1[['Education']])
df = df1

In [8]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df.drop(columns=['Personal Loan']), df['Personal Loan'], 
                                                    test_size=0.2, random_state=42)

In [36]:
from sklearn import svm
clf_svm = svm.SVC(kernel='poly')
clf_svm.fit(X_train, y_train)
y_pred = clf_svm.predict(X_test)
clf_svm.score(X_test, y_test) * 100

98.2

In [37]:
clf_svm.support_vectors_

array([[0.20454545, 0.19565217, 0.48148148, ..., 1.        , 1.        ,
        1.        ],
       [0.72727273, 0.73913043, 0.43981481, ..., 0.        , 0.        ,
        0.        ],
       [0.65909091, 0.65217391, 0.48611111, ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.40909091, 0.41304348, 0.49074074, ..., 0.        , 0.        ,
        1.        ],
       [0.90909091, 0.89130435, 0.34722222, ..., 0.        , 0.        ,
        0.        ],
       [0.36363636, 0.32608696, 0.60185185, ..., 0.        , 1.        ,
        0.        ]])

In [38]:
from sklearn import metrics
print(metrics.confusion_matrix(y_test, y_pred, labels=[0, 1]))
print("="*60)
print(metrics.classification_report(y_test, y_pred, labels=[0, 1]))

[[893   2]
 [ 16  89]]
              precision    recall  f1-score   support

           0       0.98      1.00      0.99       895
           1       0.98      0.85      0.91       105

    accuracy                           0.98      1000
   macro avg       0.98      0.92      0.95      1000
weighted avg       0.98      0.98      0.98      1000



##Conclusion<br>
SVM is one of the most popular Supervised Learning algorithms, which is used for Classification as well as Regression problems.The goal of the SVM algorithm is to create the best line or decision boundary that can segregate n-dimensional space into classes so that we can easily put the new data point in the correct category in the future.<br>
From the analysis of confusion matrix, I found that accurracy is 98%<br>
893 are True Positive, 89 are False Negative. 2 are False Positive whereas 16 are True Negative. On trying all the kernel methods, I found 'poly' as best kernel.  